Upgrade tests to use JUnit5

madmike200590 opened the pull request:
All tests were switched from JUnit version 4 to 5 (aka junit jupiter) in order to address issues with time-intensive tests running into timeouts in certain configurations. JUnit 5 offers a feature to set timeouts programmatically, i.e. flexibly according to the respective config.

The following changes were made to account for the (rather large) changes introduced in JUnit 5

### Parameterized Tests
The inheritance-based architecture for parameterization built around `AbstractSolverTests` and JUnit 4's `Parameterized` runner had to be changed since test parameterization in JUnit 5 is handled using method-level annotations, see [JUnit 5 docs](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests).

Test configs (i.e. parameter sets) for all classes previoulsy extending `AbstractSolverTests` are now supplied via the static factory methods in `at.ac.tuwien.kr.alpha.solver.RegressionTestConfigProvider`. In order to not have to put the `@ParameterizedTest` and `@MethodSource` annotations on every method parameterized using this mechanism, I introduced a new meta-annotation `@RegressionTest` that bundles the two and has the correct supplier method for parameters fixed. Note that there is a special form of this annotation, `@AggregateRegressionTest` for use in `AggregateTests` that supplies configs both with and without sorting-grid based aggregate encodings enabled. (Default for standard `RegressionTest`s is to use sorting-grid only)

### Other changes due to JUnit 5
These are a few smaller things I stumbled across during migration which I thought would be useful to document here:
* `@Test(expected = SomeThrowable.class)` was replaced with the method call `assertThrows(Throwable, Executable)`, where the `Executable` is any lambda without params or return value.
* Timeouts are now set programmatically in a fashion similar to the mechanism for exceptions. I created `TestUtils#runWithTimeout` as a convenient shorthand.
* `@Ignore` is now called `@Disabled`.
* `assertEquals` and other `assertSomething`s have a new parameter order in the version with custom error message, e.g. the junit4 method call `assertEquals("Not equal", 1, 2)` would be `assertEquals(1, 2, "Not equal")` in junit5.
* `@Before` has a new name, i.e. setup methods run before each test are now annotated with `@BeforeEach` (there is also `@BeforeClass`).

AntoniusW commented on build.gradle:
```diff
@@ -47,7 +47,9 @@ dependencies {
 	implementation group: 'org.apache.poi',     name: 'poi-ooxml',            version: '4.1.1'
 
-	testImplementation group: 'junit', name: 'junit', version: '4.12'
+	//testImplementation group: 'junit', name: 'junit', version: '4.12'
```
```suggestion
```

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/MainTest.java:
```diff
@@ -27,42 +27,52 @@
+import org.junit.jupiter.params.provider.MethodSource;
 
-@RunWith(Parameterized.class)
+//@RunWith(Parameterized.class)
```
```suggestion
```

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/MainTest.java:
```diff
@@ -27,42 +27,52 @@
-						{{"-DebugEnableInternalChecks", "-g", "naive", "-s", "default", "-n", "1", "-str", INPUT }},
-						{{"-g", "naive", "-s", "default", "-r", "naive", "-e", "1119654162577372", "--numAS", "1", "-str", INPUT }}});
-	}
+//	@Parameters
+//	public static Iterable<? extends String[][]> data() {
+//		return Arrays.asList(
+//				new String[][][] { { { "-DebugEnableInternalChecks", "-g", "naive", "-s", "default", "-e", "1119654162577372", "-n", "20", "-str", INPUT } },
+//						{ { "-DebugEnableInternalChecks", "-g", "naive", "-s", "default", "-n", "0", "-str", INPUT } },
+//						{ { "-DebugEnableInternalChecks", "-g", "naive", "-s", "default", "-n", "1", "-str", INPUT } },
+//						{ { "-g", "naive", "-s", "default", "-r", "naive", "-e", "1119654162577372", "--numAS", "1", "-str", INPUT } } });
+//	}
```
```suggestion
```

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/MainTest.java:
```diff
@@ -27,42 +27,52 @@
 
-	@Parameter
-	public String[] argv;
+//	@Parameter
+//	public String[] argv;
```
```suggestion
```

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/solver/HeadBodyTransformationTests.java:
```diff
@@ -50,119 +49,141 @@
-	}
+	private static final int DEBUG_TIMEOUT_FACTOR = 5;
+	
+//	@Before
+//	public void printSolverName() {
+//		System.out.println(solverName);
+//	}
```
```suggestion
```

AntoniusW commented on src/main/java/at/ac/tuwien/kr/alpha/solver/DefaultSolver.java:
```diff
@@ -103,6 +103,7 @@ public DefaultSolver(AtomStore atomStore, Grounder grounder, NoGoodStore store,
 		this.assignment = assignment;
 		this.store = store;
 		this.choiceManager = new ChoiceManager(assignment, store);
+		this.choiceManager.setChecksEnabled(config.isDebugInternalChecks());
```
I guess this also fixes #272 ?

madmike200590 replied:
yes, it does. PR updated accordingly.

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/solver/RegressionTest.java:
```diff
@@ -0,0 +1,17 @@
+@Retention(RetentionPolicy.RUNTIME)
+@ParameterizedTest
+@MethodSource("at.ac.tuwien.kr.alpha.solver.RegressionTestConfigProvider#provideConfigs")
+public @interface RegressionTest {
```
I would like to have a discussion about the term "regression test", because a regression --- at least to my understanding --- is a re-introduced and re-appearing bug that was already solved. Some of our tests come from bugs, so the term regression test would be suitable for them. However, most of our tests are more generally testing some functionality. So, I think the more general term of `FunctionalTest` is more fitting here (also see https://en.wikipedia.org/wiki/Functional_testing which describes regression testing as one kind of functional testing).

So, my opinion is that most stuff that is now named some kind of "regression" test would better be named "functional" test. We could then have `FunctionalTest`, `FunctionalTestConfig`, `FunctionalTestConfigProvider` and `@FunctionalTest` as annotation.
Note that some stuff might still be named with regression, namely if we really want to indicate what test cases come from regressions. An example of such would be the `DefaultSolverTest`.

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/solver/RegressionTestConfig.java:
```diff
@@ -0,0 +1,99 @@
+		return this.evaluateStratifiedPart;
+	}
+
+	public boolean isEncodeAggregatesUsingSortingGrid() {
+		return this.encodeAggregatesUsingSortingGrid;
+	}
```
IntelliJ claims that this method is never used! This may be due to an oversight in `TestUtils#regressionTestConfigToSystemConfig`which builds a `SystemConfig` from a `RegressionTestConfig` and forgets about the grid type for the aggregates. It probably should set `SystemConfig.useNormalizationGrid` to `RegressionTestConfig.isEncodeAggregatesUsingSortingGrid();`.
To my understanding, the current code only tests the default aggregate encoding grid now.

madmike200590 replied:
Thanks for catching this! It was an oversight on my part and exactly what you outlined above. Fixed it now.

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/test/util/TestUtils.java:
```diff
@@ -95,4 +117,75 @@ public static Atom basicAtomWithSymbolicTerms(String predicate, String... consta
+		retVal.setDebugInternalChecks(cfg.isDebugChecks());
+		retVal.setEvaluateStratifiedPart(cfg.isEvaluateStratifiedPart());
+		retVal.setGrounderAccumulatorEnabled(cfg.isDisableInstanceRemoval());
+		retVal.setUseNormalizationGrid(!cfg.isEncodeAggregatesUsingSortingGrid());
```
Maybe a topic for another discussion (especially since the aggregate rewriting is not yet merged): why is there the boolean negation `!` needed? When tracing this setting, I saw three different names used for this (counting grid, sorting grid, normalization grid) and sometimes the one was the other but inverted. This should be cleaned up (but maybe not here, but in the one for the aggregate rewriting).
I just would like to be sure that the negation here is correct.

AntoniusW reviewed (changes requested):
Lots of changes and many good ones. Since this PR touches a lot of old tests, I would like to leave the code a little cleaner than it was previously, hence my many comments.

Thanks for investing time in this!

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/SolverTests.java:
```diff
@@ -76,8 +71,8 @@ public int compareTo(Thingy o) {
-	@Test
-	public void testObjectProgram() throws IOException {
+	@RegressionTest
+	public void testObjectProgram(RegressionTestConfig cfg) throws IOException {
```
Thanks! These seem to be leftovers from way back since they weren't even required before the JUnit upgrade. I removed the `throws` statements in question.

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/SolverTests.java:
```diff
@@ -90,60 +85,66 @@ public void testObjectProgram() throws IOException {
-	public void testFactsOnlyProgram() throws IOException {
-		assertAnswerSet(
+	@RegressionTest
+	public void testFactsOnlyProgram(RegressionTestConfig cfg) throws IOException {
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/SolverTests.java:
```diff
@@ -90,60 +85,66 @@ public void testObjectProgram() throws IOException {
-	public void testSimpleRule() throws Exception {
-		assertAnswerSet(
+	@RegressionTest
+	public void testSimpleRule(RegressionTestConfig cfg) throws Exception {
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/AnswerSetToXlsxWriterTest.java:
```diff
@@ -28,9 +30,9 @@ public void writeAnswerSetFilesTest() throws IOException {
 		File tmpDirFile = tmpDir.toFile();
 		File[] generatedFiles = tmpDirFile.listFiles();
-		Assert.assertEquals(generatedFiles.length, 1);
+		assertEquals(generatedFiles.length, 1);
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/test/util/TestUtils.java:
```diff
@@ -95,4 +117,74 @@ public static Atom basicAtomWithSymbolicTerms(String predicate, String... consta
+	}
+	
+	public static void runWithTimeout(Executable action, long timeoutMillis) {
+		assertTimeout(Duration.ofMillis(timeoutMillis), action);
```
Thanks, that's something I didn't think of. I changed `TestUtils#runWithTimeout` to use `assertTimeoutPreemptively`.

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/heuristics/BranchingHeuristicFactoryTest.java:
```diff
@@ -56,16 +61,16 @@ public void setUp() {
 	public void testChainedHeuristicWithReplay() {
 		HeuristicsConfigurationBuilder builder = new HeuristicsConfigurationBuilder().setHeuristic(BranchingHeuristicFactory.Heuristic.VSIDS).setReplayChoices(Arrays.asList(1, 2, 3));
-		BranchingHeuristic branchingHeuristic = factory.getInstance(builder.build(), null, null, choiceManager, null);
+		BranchingHeuristic branchingHeuristic = BranchingHeuristicFactory.getInstance(builder.build(), null, null, choiceManager, null);
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/ThreeColouringWheelTest.java:
```diff
@@ -49,42 +51,51 @@
+	@RegressionTest
+	public void testN4(RegressionTestConfig cfg) throws IOException {
+		long timeout = 1000L;
+		runWithTimeout(() -> testThreeColouring(4, cfg), cfg.isDebugChecks() ? timeout * DEBUG_TIMEOUT_FACTOR : timeout);
```
I agree - due to the complexity of the expression, I also produced a few copy-paste errors when passing timeout params that I only found now... I changed `TestUtils#runWithTimeout` to take a `RegressionTestConfig`, a timeout value, and the debug scaling factor as params and do everything else internally.

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/ThreeColouringWheelTest.java:
```diff
@@ -49,42 +51,51 @@
+	@RegressionTest
+	public void testN5(RegressionTestConfig cfg) throws IOException {
+		long timeout = 1000L;
+		runWithTimeout(() -> testThreeColouring(5, cfg), cfg.isDebugChecks() ? timeout * DEBUG_TIMEOUT_FACTOR : timeout);
```
fixed in `TestUtils#runWithTimeout`

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/ThreeColouringTestWithRandom.java:
```diff
@@ -51,114 +53,135 @@
+	@RegressionTest
+	public void testN3(RegressionTestConfig cfg) throws IOException {
+		long timeout = 3000L;
+		runWithTimeout(() -> testThreeColouring(3, false, 0, cfg), cfg.isDebugChecks() ? timeout * DEBUG_TIMEOUT_FACTOR : timeout);
```
fixed in `TestUtils#runWithTimeout`

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/ThreeColouringRandomGraphTest.java:
```diff
@@ -43,49 +45,59 @@
+	@RegressionTest
+	public void testV3E3(RegressionTestConfig cfg) throws IOException {
+		long timeout = 1000L;
+		runWithTimeout(() -> testThreeColouring(3, 3, cfg), cfg.isDebugChecks() ? timeout * DEBUG_TIMEOUT_FACTOR : timeout);
```
fixed in `TestUtils#runWithTimeout`

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/MainTest.java:
```diff
@@ -27,42 +27,41 @@
+import org.junit.jupiter.params.provider.MethodSource;
 
-@RunWith(Parameterized.class)
 public class MainTest {
```
Not sure.. `Main` is, by convention, the main commandline entry point, therefore a name like `TestCommandlineMainEntryPoint` feels kinda redundant to me

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/MainTest.java:
```diff
@@ -27,42 +27,41 @@
-	public String[] argv;
+
+	
+	public static Stream<Arguments> provideArgs() {
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/MainTest.java:
```diff
@@ -27,42 +27,41 @@
-	@Test
-	public void test() {
+	@ParameterizedTest
+	@MethodSource("provideArgs")
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/MainTest.java:
```diff
@@ -27,42 +27,41 @@
+	}
 
 	/**
 	 * Temporarily redirects System.err and System.out while running the solver from the main entry point with the
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/MainTest.java:
```diff
@@ -71,8 +70,9 @@ public void test() {
-	@Test
-	public void filterTest() {
+	@ParameterizedTest
+	@MethodSource("provideArgs")
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/TrailAssignmentTest.java:
```diff
@@ -59,20 +61,24 @@ public TrailAssignmentTest() {
 
-	@Test(expected = IllegalArgumentException.class)
+	@Test
 	public void assign() throws Exception {
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/TrailAssignmentTest.java:
```diff
@@ -59,20 +61,24 @@ public TrailAssignmentTest() {
 
-	@Test(expected = IllegalArgumentException.class)
+	@Test
 	public void negativeAtomThrows() throws Exception {
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/UtilTest.java:
```diff
@@ -1,20 +1,29 @@
-	@Test(expected = RuntimeException.class)
+	
+	@Test
 	public void oops() throws Exception {
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/grounder/transformation/StratifiedEvaluationRegressionTest.java:
```diff
@@ -90,37 +89,28 @@
-		String aspStr = this.aspString;
+	@ParameterizedTest
+	@MethodSource("at.ac.tuwien.kr.alpha.grounder.transformation.StratifiedEvaluationRegressionTest#params")
+	public void runTest(String aspString, Consumer<InternalProgram> programVerifier, Consumer<Set<AnswerSet>> resultVerifier) {
+		String aspStr = aspString;
 		Alpha system = new Alpha();
 		InputProgram prg = system.readProgramString(aspStr);
 		NormalProgram normal = system.normalizeProgram(prg);
 		AnalyzedProgram analyzed = AnalyzedProgram.analyzeNormalProgram(normal);
 		InternalProgram evaluated = new StratifiedEvaluation().apply(analyzed);
-		this.programVerifier.accept(evaluated);
+		programVerifier.accept(evaluated);
 		Set<AnswerSet> answerSets = system.solve(evaluated).collect(Collectors.toSet());
-		this.answerSetsVerifier.accept(answerSets);
+		resultVerifier.accept(answerSets);
```
As per our discussion, keeping it as is for now, although I agree it's maybe not an ideal structure.

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/DefaultSolverTest.java:
```diff
@@ -0,0 +1,48 @@
+
+public class DefaultSolverTest {
+	
+	// detailed reproduction test-case for github issue #239
```
See our discussion, this is a reproduction test that doesn't have anything to do with the test structure previously defined in `AbstractSolverTests` and was therefore moved to a separate class.

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/HanoiTowerTest.java:
```diff
@@ -34,71 +46,69 @@
-		testHanoiTower(1);
+	@RegressionTest
+	@Disabled("disabled to save resources during CI")
+	public void testInstance1(RegressionTestConfig cfg) throws IOException {
```
fixed

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/test/util/TestUtils.java:
```diff
@@ -95,4 +117,75 @@ public static Atom basicAtomWithSymbolicTerms(String predicate, String... consta
+		return collectRegressionTestAnswerSets(prog, cfg);
+	}
+
+	public static void assertRegressionTestAnswerSet(String program, String answerSet, RegressionTestConfig cfg) {
```
Yeah, that makes sense. Fixed it

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/solver/AggregatesTest.java:
```diff
@@ -25,115 +25,111 @@
+	@AggregateRegressionTest
+	public void testAggregate_Sum_Computed(RegressionTestConfig cfg) throws IOException {
+		// Do not run this test case with the naive solver.
+		if (cfg.getSolverName().equals("naive")) {
+			return;
+		}
```
I neglected to use the proper method from `TestUtils`, fixed.

madmike200590 commented on src/test/java/at/ac/tuwien/kr/alpha/test/util/TestUtils.java:
```diff
@@ -95,4 +117,75 @@ public static Atom basicAtomWithSymbolicTerms(String predicate, String... consta
+	
+	private static Solver buildSolverFromSystemConfig(InputProgram prog, SystemConfig cfg) {
+		AtomStore atomStore = new AtomStoreImpl();
+		NormalProgram normalProg = new NormalizeProgramTransformation(cfg.isUseNormalizationGrid()).apply(prog); // TODO we need to somehow handle aggregate configs
```
nope ;) removed it

madmike200590 commented:
@AntoniusW thanks for the detailed review! I applied your suggestions and addressed the remaining issues now, please have another look and merge if ok.

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/AnswerSetToXlsxWriterTest.java:
```diff
@@ -28,12 +30,13 @@ public void writeAnswerSetFilesTest() throws IOException {
-		AnswerSetToWorkbookMapperTest.assertWorkbookMatchesAnswerSet(wb, as);
-		wb.close();
+		assertEquals("alphaAnswerSet.0.xlsx", answerSetFile.getName());
+		try (Workbook wb = WorkbookFactory.create(answerSetFile)) {
+			AnswerSetToWorkbookMapperTest.assertWorkbookMatchesAnswerSet(wb, as);
+			wb.close();
+		}
```
Interesting, I didn't know about the try-with-resources statement. My IDE claims that the `.close()` is redundant and can be deleted (in this and all other try-with-resources blocks).
```suggestion
		try (Workbook wb = WorkbookFactory.create(answerSetFile)) {
			AnswerSetToWorkbookMapperTest.assertWorkbookMatchesAnswerSet(wb, as);
		}
```

madmike200590 replied:
Thanks, I overlooked that - removed obsolete close statements now

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/AnswerSetToXlsxWriterTest.java:
```diff
@@ -45,17 +48,18 @@ public void writeUnsatTest() throws IOException {
+			assertNotNull(cell);
+			String cellValue = cell.getStringCellValue();
+			assertEquals("Input is unsatisfiable - No answer sets!", cellValue);
+			wb.close();
```
```suggestion
```

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/api/mapper/AnswerSetToWorkbookMapperTest.java:
```diff
@@ -26,12 +29,13 @@
+			assertNotNull(wb.getSheet("bla_2"));
+			assertNotNull(wb.getSheet("foo_1"));
+			assertNotNull(wb.getSheet("complex_3"));
+			wb.close();
```
```suggestion
```

AntoniusW commented on src/test/java/at/ac/tuwien/kr/alpha/test/util/TestUtils.java:
```diff
@@ -95,4 +117,78 @@ public static Atom basicAtomWithSymbolicTerms(String predicate, String... consta
 		return new BasicAtom(pred, trms);
 	}
 
+	private static SystemConfig regressionTestConfigToSystemConfig(RegressionTestConfig cfg) {
```
If I remember correctly, then we came to the conclusion to move this method to `RegressionTestConfig`. Was there a reason not to move it?

AntoniusW commented:
Looks all good. There are two open points and after deciding how to handle them, you @madmike200590, can merge this PR. One point is unnecessary `wb.close();` inside try-with-resources blocks, the other a method I thought we said to move to another class. Thank you again for this contribution!
